### PR TITLE
feat: new rule landmark-complementary-is-top-level

### PR DESCRIPTION
**Pull request by marcysutton**
Best practice requiring asides and complementary landmarks to be top level, in line with the ARIA Authoring Practices Guide.

Closes https://github.com/dequelabs/axe-core/issues/795

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [ ] Follows the commit message policy, appropriate for next version
- [ ] Has documentation updated, a DU ticket, or requires no documentation change
- [ ] Includes new tests, or was unnecessary
- [ ] Code is reviewed for security by: << Name here >>

### CODE REVIEW

File: lib/rules/landmark-has-body-context.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 const nativeScopeFilter = 'article, aside, main, nav, section';
 
 // Filter elements that, within certain contexts, don't map their role.
-// e.g. a <footer> inside a <main> is not a banner, but in the <body> context it is
+// e.g. a <header> inside a <main> is not a banner, but in the <body> context it is
 return (
 	node.hasAttribute('role') ||
```
Comment (jkodu) [inline lib/rules/landmark-has-body-context.js]: Could this be extracted to `"selector": "[role]"`?

File: test/checks/keyboard/landmark-is-top-level.js [hunk 2]
```diff
@@ -23,6 +23,16 @@ describe('landmark-is-top-level', function() {
 		assert.deepEqual(checkContext._data, { role: 'main' });
 	});
 
+	it('should return false if the complementary landmark is in another landmark', function() {
+		var mainLandmark = document.createElement('main');
+		var aside = document.createElement('div');
+		aside.setAttribute('role', 'complementary');
+		aside.appendChild(mainLandmark);
+		fixture.appendChild(aside);
```
Comment (jkodu) [inline test/checks/keyboard/landmark-is-top-level.js]: Nitpick: I know, there are multiple tests written with DOM api like above using `createElement`, `appendChild` etc. I found that using the `testutils.fixtureSetup()` function to be helpful for this, and saves us a few lines and also makes it easy to read.

File: test/checks/keyboard/landmark-is-top-level.js [hunk 3]
```diff
@@ -23,6 +23,16 @@ describe('landmark-is-top-level', function() {
 		assert.deepEqual(checkContext._data, { role: 'main' });
 	});
 
+	it('should return false if the complementary landmark is in another landmark', function() {
```
Comment (jkodu) [inline test/checks/keyboard/landmark-is-top-level.js]: Also, just had a thought, what happens if the first `header` element after `body` element happens with in `shadowDOM`?

Question being, do we need to consider `shadowDOM` for this rule?

Comment (marcysutton) [reply test/checks/keyboard/landmark-is-top-level.js]: The check already has Shadow DOM coverage.

**Comment by marcysutton** [comment]
@jkodu thanks for the tip. Does that require installing circleci globally on your machine?

File: lib/rules/landmark-has-body-context.js (hunk 1 above)
Comment (marcysutton) [reply lib/rules/landmark-has-body-context.js]: I ended up not using this file, and the only thing I changed in it was a comment typo. Leaving as-is.

File: test/checks/keyboard/landmark-is-top-level.js (hunk 2 above)
Comment (marcysutton) [reply test/checks/keyboard/landmark-is-top-level.js]: I've updated the tests, you're right that they were more difficult to read!

File: lib/rules/landmark-has-body-context.js (hunk 1 above)
Comment (WilcoFiers) [reply lib/rules/landmark-has-body-context.js]: I had to look it up too, but yes aside always maps to role= complementary, unlike footer and header which only map to a role if they are scoped to the body.

**Pull request merged**